src/Microsoft.ML.Ensemble/OutputCombiners/RegressionStacking.cs

```diff
@@ -19,7 +21,7 @@ namespace Microsoft.ML.Runtime.Ensemble.OutputCombiners
 {
     using TScalarPredictor = IPredictorProducing<Single>;
 
-    public sealed class RegressionStacking : BaseScalarStacking<SignatureRegressorTrainer>, IRegressionOutputCombiner, ICanSaveModel
```

Comment:
>SignatureRegressorTrainer [](start = 64, length = 25)

What are we going to do about this sort of restriction? I had suggested previously using something like `ITrainer<IRegressionPredictor>` or something like this, but I'm not sure what you thought of that idea.

Reply:
Which restriction are you referring to?  The Signature Type restriction, where we can't use `typeof(TSig)` in an attribute?

My hope is that some day we remove signature types all together.  I haven't thought of/discovered a good replacement for them yet. Maybe an approach similar to [ComponentKind](https://github.com/dotnet/machinelearning/blob/5e08fa1ea7bfb54f28ed0815cb6413e0068e6dd1/src/Microsoft.ML.Core/EntryPoints/ModuleArgs.cs#L85-L93) which is just a string "Kind".

However, that still wouldn't help in this case, because we need a different signature/kind/etc for each concrete class, but we want to define the argument on the base class.

Reply:
I meant just restricting the DI and API usage to only consider actual, in this case, regression trainers, which is one of the functions of the signatures. Currently this is done in a fairly half-hearted fashion using this `IPredictorProducing<float>` and `IPredictorProducing<VBuffer<float>>`, but if we got rid of this and replaced it with, say, `IRegressionPredictor` and `IBinaryClassificationPredictor` and `IMulticlassPredictor` (or something like that), then we could just have this be c omponent factory producing `ITrainer<IRegressionPredictor>` in this case. I think MEF should work well with that restriction as well, from a DI perspective.

---
In reply to: [210463209](https://github.com/dotnet/machinelearning/pull/681#discussion_r210463209) [](ancestors = 210463209)